### PR TITLE
feat: made a new password input component

### PR DESCRIPTION
**Pull request by NasgulNexus**

### CODE REVIEW

File: src/components/Password/Password.tsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import {Copy, CopyCheck, Eye, EyeSlash} from '@gravity-ui/icons';
+import {
+    Button,
+    CopyToClipboard,
+    CopyToClipboardStatus,
+    Icon,
+    TextInput,
+    TextInputProps,
+} from '@gravity-ui/uikit';
+import _ from 'lodash';
```
Comment (personaljs) [inline src/components/Password/Password.tsx:12]: зачем тут lodash?

File: src/components/Password/Password.tsx [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import './Password.scss';
+
+const b = block('password');
+
+export type PasswordProps = Required<Pick<TextInputProps, 'onUpdate' | 'value'>> &
+    Omit<TextInputProps, 'rightContent' | 'type'> & {
+        showCopyButton?: boolean;
+    };
+
+export const Password: React.FC<PasswordProps> = (props) => {
```
Comment (amje) [inline src/components/Password/Password.tsx:24]: `Password` -> `PasswordInput`

File: src/components/Password/Password.tsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+import {block} from '../utils/cn';
+
+import './Password.scss';
+
+const b = block('password');
+
+export type PasswordProps = Required<Pick<TextInputProps, 'onUpdate' | 'value'>> &
+    Omit<TextInputProps, 'rightContent' | 'type'> & {
+        showCopyButton?: boolean;
```
Comment (amje) [inline src/components/Password/Password.tsx:21]: `showVisibilityButton` should also be in the props

File: src/components/Password/Password.tsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+    const [hideValue, setHideValue] = React.useState(true);
+
+    const additionalRightContent = React.useMemo(() => {
+        const onClick = () => {
+            setHideValue((hideValue) => !hideValue);
+        };
+
+        return (
+            <div className={b()}>
```
Comment (amje) [inline src/components/Password/Password.tsx:35]: Root class is assigned to the extra content, not the TextInput

File: src/components/Password/Password.tsx [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+} from '@gravity-ui/uikit';
+
+import {block} from '../utils/cn';
+
+import './Password.scss';
+
+const b = block('password');
+
+export type PasswordProps = Required<Pick<TextInputProps, 'onUpdate' | 'value'>> &
+    Omit<TextInputProps, 'rightContent' | 'type'> & {
```
Comment (amje) [inline src/components/Password/Password.tsx:20]: Why omit `rightContent`? We can prepend it to buttons

**Review by amje — CHANGES_REQUESTED** [review]

File: src/components/PasswordInput/PasswordInput.tsx [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+                            </Button>
+                        )}
+                    </CopyToClipboard>
+                ) : null}
+                {showVisibilityButton ? (
+                    <Button
+                        view="flat-secondary"
+                        onClick={onClick}
+                        className={b('button')}
+                        size="s"
```
Comment (amje) [inline src/components/PasswordInput/PasswordInput.tsx:64]: We should use different sizes for each TextInput size

File: src/components/PasswordInput/PasswordInput.tsx [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+                                        state === CopyToClipboardStatus.Pending ? Copy : CopyCheck
+                                    }
+                                />
+                            </Button>
+                        )}
+                    </CopyToClipboard>
+                ) : null}
+                {showVisibilityButton ? (
+                    <Button
+                        view="flat-secondary"
```
Comment (amje) [inline src/components/PasswordInput/PasswordInput.tsx:61]: aria-label prop would be nice here with the texts: `Show password/Hide password`

File: src/components/PasswordInput/PasswordInput.tsx [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        );
+    }, [rightContent, value, showCopyButton, showVisibilityButton, hideValue]);
+
+    return (
+        <TextInput
+            {...props}
+            type={hideValue ? 'password' : 'text'}
+            rightContent={additionalRightContent}
+            autoComplete={autoComplete ? autoComplete : 'new-password'}
+            className={b('input', className)}
```
Comment (amje) [inline src/components/PasswordInput/PasswordInput.tsx:79]: That's not the input's className, you should pass it to the `controlProps` prop

**Review by amje — CHANGES_REQUESTED** [review]

File: src/components/PasswordInput/PasswordInput.tsx [hunk 9]
```diff
@@ -20,18 +15,27 @@ export type PasswordInputProps = Required<Pick<TextInputProps, 'onUpdate' | 'val
     Omit<TextInputProps, 'type'> & {
         /** Show copy button */
         showCopyButton?: boolean;
-        /** Show visibility button */
-        showVisibilityButton?: boolean;
+        /** Show reveal button */
+        showRevealButton?: boolean;
+        /** Disable tooltip. Tooltip won't be shown */
+        hasTooltip?: boolean;
```
Comment (amje) [inline src/components/PasswordInput/PasswordInput.tsx:21]: I would pick separate names here, `hasCopyTooltip` and `hasRevealTooltip`

File: src/components/PasswordInput/PasswordInput.tsx [hunk 10]
```diff
@@ -43,40 +47,47 @@ export const PasswordInput: React.FC<PasswordInputProps> = (props) => {
...
-                                        state === CopyToClipboardStatus.Pending ? Copy : CopyCheck
-                                    }
-                                />
-                            </Button>
-                        )}
-                    </CopyToClipboard>
+                    <ClipboardButton
+                        text={value}
+                        hasTooltip={hasTooltip}
+                        size={16}
```
Comment (amje) [inline src/components/PasswordInput/PasswordInput.tsx:53]: We should map input size to button size

File: src/components/PasswordInput/PasswordInput.tsx [hunk 11]
```diff
@@ -43,40 +47,47 @@ export const PasswordInput: React.FC<PasswordInputProps> = (props) => {
...
+                        content={
+                            hideValue ? i18n('label_show-password') : i18n('label_hide-password')
+                        }
                     >
-                        <Icon data={hideValue ? Eye : EyeSlash} size={14} />
-                    </Button>
+                        <Button
+                            view="flat-secondary"
+                            onClick={onClick}
+                            size={size}
```
Comment (amje) [inline src/components/PasswordInput/PasswordInput.tsx:67]: TextInput size !== Button size, take a look at clear button, maybe reuse this [logic](https://github.com/gravity-ui/uikit/blob/e9aff632b924307e7700e76855ba15abc905be3b/src/components/controls/common/ClearButton/ClearButton.tsx#L23) here as well

File: src/components/PasswordInput/PasswordInput.tsx [hunk 12]
```diff
@@ -43,40 +47,47 @@ export const PasswordInput: React.FC<PasswordInputProps> = (props) => {
...
 
     return (
         <TextInput
             {...props}
             type={hideValue ? 'password' : 'text'}
             rightContent={additionalRightContent}
             autoComplete={autoComplete ? autoComplete : 'new-password'}
-            className={b('input', className)}
+            controlProps={{
+                className: b(),
```
Comment (amje) [inline src/components/PasswordInput/PasswordInput.tsx:89]: Now it's an opposite, you're passing root className (`b()`) to the control node

**Review by amje — CHANGES_REQUESTED** [review]

File: src/components/PasswordInput/__stories__/PasswordInput.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+            counter += 1;
+        }
+
+        setValue(randomValue);
+    }, []);
+
+    return (
+        <div className={b()}>
+            <PasswordInput {...props} onUpdate={setValue} value={value} />
+            <Button onClick={generateRandomValue} className={b('button-generate-random-value')}>
```
Comment (personaljs) [inline src/components/PasswordInput/__stories__/PasswordInput.stories.tsx:55]: давай добавим showCopyButton={true} showRevealButton={true}

File: src/components/PasswordInput/PasswordInput.tsx [hunk 14]
```diff
@@ -86,7 +102,8 @@ export const PasswordInput: React.FC<PasswordInputProps> = (props) => {
             rightContent={additionalRightContent}
             autoComplete={autoComplete ? autoComplete : 'new-password'}
             controlProps={{
-                className: b(),
+                className: b('input-control', controlProps?.className),
+                ...controlProps,
```
Comment (amje) [inline src/components/PasswordInput/PasswordInput.tsx:106]: Spread should be before the className, otherwise it would be overwritten by the controlProps

**Review by amje — CHANGES_REQUESTED** [review]

**Review by amje — APPROVED** [review]

**Pull request merged**